### ISSUE TITLE
DL4J: fit(DataSetIterator) with pretrain(false) - no warning/error

### ISSUE THREAD

**Issue by devlop16** [deeplearning4j/deeplearning4j#6228]
If you try this code in beta1 or beta2 or current snapshot:

```
public void TestSimpleVAE(){
     MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
            .seed(1234)
            .updater(new RmsProp(1e-2)).biasUpdater(new RmsProp(1e-2))
            .weightInit(WeightInit.XAVIER)
            .l2(1e-4)
            .list()
            .layer(0, new VariationalAutoencoder.Builder()
                .activation(Activation.LEAKYRELU)
                .encoderLayerSizes(256, 256)        //2 encoder layers, each of size 256
                .decoderLayerSizes(256, 256)        //2 decoder layers, each of size 256
                .pzxActivationFunction(Activation.IDENTITY)  //p(z|data) activation function
                .biasInit(0.01).visibleBiasInit(0.01)
                .reconstructionDistribution(new BernoulliReconstructionDistribution(Activation.SIGMOID.getActivationFunction()))     //Bernoulli distribution for p(data|z) (binary or 0 to 1 data only)
                .nIn(4)                    
                .nOut(2)                            //Size of the latent variable space: p(z|x). 2 dimensions here for plotting, use more in general
                .build())
            .pretrain(true).backprop(false).build();

        MultiLayerNetwork net = new MultiLayerNetwork(conf);
        net.init();
        int batchsize=10;
        Random r= new Random();
        INDArray inputs = Nd4j.create(10,4);
          for(int a=0;a<batchsize;a++){
                INDArray data0;    
                double[] data1= new double[4];
                for(int d=0;d<data1.length;d++)
                    data1[d]=r.nextDouble();
                data0= Nd4j.create(data1);    
                inputs.putRow(a, data0);                
                }
                    
       DataSet d2= new DataSet(inputs,inputs);  
       net.fit(d2);   
       System.out.println(net.paramTable());
}
```

The output of the bias layers in beta2 is always:
0_e0b=[[         0,         0,         0,         0.....
0_e0b=[[         0,         0,         0,         0......
0_pZXMeanb=[[   0,    0]]
0_pZXLogStd2b=[[         0,         0]]
0_d1b=[[         0,         0,         0,         0,         0,         0,    ......
 0_pXZb=[[         0,         0,         0,         0]]

but if you do it in alpha version or earlier the output is something like:

0_e0b=[[   -0,0024,   -0,0033,    0,0045,    0,0045.....
0_e1b=[[   -0,0029,   -0,0044,    0,0045,    0,0043, ....
0_pZXMeanb=[[   -0,0045,    0,0045]]
0_pZXLogStd2b=[[    0,0045,    0,0045]],
0_d1b=[[   -0,0016,   -0,0045,    0,0045,   -0,0045,  ....
0_pXZb=[[    0,0045,    0,0045,    0,0045,    0,0045]

This is tested on CPU but clearly shows beta versions is not dealing correctly with bias

**Comment by AlexDBlack** [comment]
Are you sure it's not just a difference in the formatting of the INDArrays - i.e., truncated/limited precision?
You can use NDArrayStrings to customize the number of decimal places.
Or, use INDArray.getDouble(...) to check also.

**Comment by devlop16** [comment]
GetDouble returns also Zeros, and it is the same code just change in the POMs the beta2 or alpha version and test by yourself

**Comment by devlop16** [comment]
In fact it ruins all the accuracy of the VAE since over training it keept the biases to Zeros, making unsuable beta2 for VAEs atm...

**Comment by devlop16** [comment]
Also tested other functions for biasUpdater and changing other parameters of the net with the same result.

**Comment by devlop16** [comment]
Tested on several machines with the same result

**Comment by AlexDBlack** [comment]
OK, so looks like this is due to a behaviour API change:
https://deeplearning4j.org/release-notes.html#deeplearning4j-api-changes-transition-guide-091-to-100-alpha

> Behaviour change: fit(DataSetIterator) and similar methods no longer perform layerwise pretraining followed by backprop - only backprop is performed in these methods. For pretraining, use pretrain(DataSetIterator) and pretrain(MultiDataSetIterator) methods

If you use pretrainLayer(0, DataSetIterator) it works as expected.

I'll leave this open until we improve the usability for this - users should get a warning or error for this.

**Comment by devlop16** [comment]
Thanks @AlexDBlack it fixes it,  
in the previous example you need to add   net.pretrainLayer(0, d2.getFeatures());      for it to work before the fit.
I think the misleading part is that if you set  .pretrain(true) you expect it to be done

**Comment by AlexDBlack** [comment]
I agree it's misleading currently (hence why it'll now throw an exception and tell you to use one of the pretrain methods).
The reason we decided to change the behaviour is that the old behaviour doesn't really make sense:
i.e., you don't want to do (pretrain layer 0, pretrain layer 1, backprop full net) per batch or per epoch, which is what it did.
Instead, the common/sensible approach to layerwise pretraining is (N epochs pretrain layer 0, N epochs pretrain layer 1, M epochs backprop full net), where N and M may be different.

**Comment by devlop16** [comment]
Thanks Alex, in this case since backprop is false, the fit is not need only the pretrain?

**Comment by AlexDBlack** [comment]
That's correct. No fit call required, only pretrain/pretrainLayer.
